### PR TITLE
feat(android): changing managed config restarts TunnelService

### LINKED ISSUE #4040: Listen for MDM config changes and rebuild TunnelService VPN
Need to listen for changes to the following fields in the AppRestrictions within TunnelService and then restart the Connlib session if they change:

```
token
allowedApplications
disallowedApplications
```

Unfortunately we can't call `BuildVpnService` instead because that generates a new `fd` which connlib won't know how to receive. So we should simply call `session.disconnect` and then `Session.connect` again.

### PR DESCRIPTION
**Pull request by jasonboukheir**
Fixes #4040

Depends on #4180

### CODE REVIEW

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt [hunk 1]
```diff
@@ -175,11 +175,16 @@ class TunnelService : VpnService() {
         startId: Int,
     ): Int {
         Log.d(TAG, "onStartCommand")
-        if (intent?.getBooleanExtra("startedByUser", false) == true) {
-            startedByUser = true
-        }
+        when (intent?.action) {
+            "dev.firezone.android.tunnel.DISCONNECT" -> disconnect()
```
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt]: typo?

```suggestion
            "dev.firezone.android.tunnel.action.DISCONNECT" -> disconnect()
```

**Comment by jamilbk** [comment]
@jasonboukheir Another approach that might work better is simply calling `session.disconnect()` and then `connect()` again with the same parameters. It might take a little refactoring to make DRY up the calls, but that would avoid any system side effects and probably be much faster.

**Comment by jasonboukheir** [comment]
So there is one hiccup here... https://developer.android.com/work/managed-configurations#listen-configuration

> Note: The [ACTION_APPLICATION_RESTRICTIONS_CHANGED](https://developer.android.com/reference/android/content/Intent#ACTION_APPLICATION_RESTRICTIONS_CHANGED) intent is sent only to listeners that are dynamically registered, not to listeners that are declared in the app manifest.

This means that for the receiver to be listening, we need the service up and not destroyed. However, on disconnect, I've seen the service sometimes being destroyed (other times, this works just fine). When the service is destroyed, the receiver isn't receiving changes, so changes in the token won't restart the service.

I'm going to test some more and see if Always-On VPN periodically restarts the service. If not, I was able to get the service to restart by turning off/on Always-On VPN setting in MDM.

**Comment by jasonboukheir** [comment]
When the tunnel becomes disconnected, the service is destroyed, and it can no longer listen for callbacks from changes to MDM.

The admin can force the service to be recreated by any of the following methods:
1. Restarting the device
2. Selecting Firezone as the Always-On VPN again in the MDM software

Outside of the tunnel being connected, I was able to verify that this feature works.

**Comment by jamilbk** [comment]
@jasonboukheir Could you rebase? I think you have a conflict

**Review by jamilbk — APPROVED** [review]
LGTM, just needs a rebase and then I'll send to customer for further testing.

**Pull request merged**